File: crates/snipx-core/src/snippet.rs
use rowan::NodeOrToken;

use crate::syntax::{SyntaxKind, SyntaxNode};

#[derive(Debug, Clone, PartialEq)]
pub enum SnippetPart {
    /// Raw unquoted body text, matched verbatim.
    Text(String),
    /// A quoted run. `raw` includes the delimiters and undecoded escapes;
    /// `decoded` strips the delimiters and decodes only `\"` -> `"`.
    Quoted {
        raw: String,
        decoded: String,
        terminated: bool,
    },
    /// `{...}` capture; `text` is the raw inner text.
    Capture { text: String, terminated: bool },
    /// An unquoted, top-level `..` in a range snippet.
    RangeSeparator,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Cardinality {
    ExactlyOne,
    OneOrMore,
    ZeroOrMore,
    ZeroOrOne,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SnippetValue {
    /// Trimmed source syntax as today (`[Alice]+`), `~` already stripped
    /// by the caller for text-span snippets. Feeds JSON output and
    /// diagnostic messages unchanged.
    pub source: String,
    pub parts: Vec<SnippetPart>,
    pub cardinality: Cardinality,
    /// False when the closing `]` is missing.
    pub terminated: bool,
}

impl SnippetValue {
    /// `node` must be a `Snippet` or `RangeSnippet` node.
    pub fn from_node(node: &SyntaxNode, source: String) -> SnippetValue {
        let is_range = node.kind() == SyntaxKind::RangeSnippet;
        let mut parts = Vec::new();
        let mut cardinality = Cardinality::ExactlyOne;
        let mut terminated = false;

        for element in node.children_with_tokens() {
            match element {
                NodeOrToken::Token(token) => match token.kind() {
                    SyntaxKind::LBrack => {}
                    SyntaxKind::RBrack => terminated = true,
                    SyntaxKind::Dot => parts.push(SnippetPart::RangeSeparator),
                    SyntaxKind::Text => {
                        // In a range snippet, split text on ".." to create separators
                        if is_range {
                            split_on_range_separator(token.text(), &mut parts);
                        } else {
                            parts.push(SnippetPart::Text(token.text().to_owned()));
                        }
                    }
                    _ => parts.push(SnippetPart::Text(token.text().to_owned())),
                },
                NodeOrToken::Node(child) => match child.kind() {
                    SyntaxKind::QuotedSnippetPart => parts.push(quoted_part(&child)),
                    SyntaxKind::Capture => parts.push(capture_part(&child)),
                    SyntaxKind::Quantifier => {
                        cardinality = match child.to_string().as_str() {
                            "+" => Cardinality::OneOrMore,
                            "*" => Cardinality::ZeroOrMore,
                            "?" => Cardinality::ZeroOrOne,
                            _ => Cardinality::ExactlyOne,
                        };
                    }
                    // Invalid captures (second capture, capture in a range)
                    // are wrapped in an Error node by the parser; surface
                    // them so the matcher reports the same InvalidSnippet
                    // errors the string lexer used to.
                    SyntaxKind::Error => match child
                        .descendants()
                        .find(|inner| inner.kind() == SyntaxKind::Capture)
                    {
                        Some(capture) => parts.push(capture_part(&capture)),
                        None => parts.push(SnippetPart::Text(child.to_string())),
                    },
                    _ => parts.push(SnippetPart::Text(child.to_string())),
                },
            }
        }

        SnippetValue {
            source,
            parts,
            cardinality,
            terminated,
        }
    }
}

/// Split text on ".." in a range snippet, creating RangeSeparator parts.
fn split_on_range_separator(text: &str, parts: &mut Vec<SnippetPart>) {
    let mut remaining = text;
    while let Some(pos) = remaining.find("..") {
        if pos > 0 {
            parts.push(SnippetPart::Text(remaining[..pos].to_owned()));
        }
        parts.push(SnippetPart::RangeSeparator);
        remaining = &remaining[pos + 2..];
    }
    if !remaining.is_empty() {
        parts.push(SnippetPart::Text(remaining.to_owned()));
    }
}

/// Snippet quoting is very literal: only the quote delimiter itself is
/// escaped, so decoding maps `\"` to `"` and nothing else.
fn quoted_part(node: &SyntaxNode) -> SnippetPart {
    let raw = node.to_string();
    let mut quotes = 0usize;
    let mut content = String::new();
    for element in node.descendants_with_tokens() {
        if let NodeOrToken::Token(token) = element {
            match token.kind() {
                SyntaxKind::Quote => quotes += 1,
                SyntaxKind::Text => content.push_str(token.text()),
                _ => {}
            }
        }
    }
    SnippetPart::Quoted {
        raw,
        decoded: content.replace("\\\"", "\""),
        terminated: quotes >= 2,
    }
}

fn capture_part(node: &SyntaxNode) -> SnippetPart {
    let mut text = String::new();
    let mut terminated = false;
    for element in node.children_with_tokens() {
        if let NodeOrToken::Token(token) = element {
            match token.kind() {
                SyntaxKind::Text => text.push_str(token.text()),
                SyntaxKind::RBrace => terminated = true,
                _ => {}
            }
        }
    }
    SnippetPart::Capture { text, terminated }
}

File: packages/snipx-ts/src/snippet.ts
/** Mirror of crates/snipx-core/src/snippet.rs. */

import type { SyntaxNode } from "./syntax.js";

export type SnippetPart =
  | { type: "text"; text: string }
  | { type: "quoted"; raw: string; decoded: string; terminated: boolean }
  | { type: "capture"; text: string; terminated: boolean }
  | { type: "rangeSeparator" };

export type Cardinality = "exactlyOne" | "oneOrMore" | "zeroOrMore" | "zeroOrOne";

export interface SnippetValue {
  /** Trimmed source syntax (`[Alice]+`), `~` already stripped by the caller. */
  source: string;
  parts: SnippetPart[];
  cardinality: Cardinality;
  /** False when the closing `]` is missing. */
  terminated: boolean;
}

/** `node` must be a Snippet or RangeSnippet node. */
export function snippetValueFromNode(node: SyntaxNode, source: string): SnippetValue {
  const parts: SnippetPart[] = [];
  let cardinality: Cardinality = "exactlyOne";
  let terminated = false;

  for (const element of node.children) {
    if (element.type === "token") {
      switch (element.kind) {
        case "LBrack":
          break;
        case "RBrack":
          terminated = true;
          break;
        case "Dot":
          parts.push({ type: "rangeSeparator" });
          break;
        default:
          parts.push({ type: "text", text: element.text });
          break;
      }
    } else {
      switch (element.kind) {
        case "QuotedSnippetPart":
          parts.push(quotedPart(element));
          break;
        case "Capture":
          parts.push(capturePart(element));
          break;
        case "Quantifier": {
          const text = element.toText();
          cardinality =
            text === "+"
              ? "oneOrMore"
              : text === "*"
                ? "zeroOrMore"
                : text === "?"
                  ? "zeroOrOne"
                  : "exactlyOne";
          break;
        }
        // Invalid captures (second capture, capture in a range) are
        // wrapped in an Error node by the parser; surface them so the
        // matcher reports the same InvalidSnippet errors.
        case "Error": {
          let capture: SyntaxNode | null = null;
          for (const inner of element.descendants()) {
            if (inner.kind === "Capture") {
              capture = inner;
              break;
            }
          }
          if (capture !== null) {
            parts.push(capturePart(capture));
          } else {
            parts.push({ type: "text", text: element.toText() });
          }
          break;
        }
        default:
          parts.push({ type: "text", text: element.toText() });
          break;
      }
    }
  }

  return { source, parts, cardinality, terminated };
}

/**
 * Snippet quoting is very literal: only the quote delimiter itself is
 * escaped, so decoding maps `\"` to `"` and nothing else.
 */
function quotedPart(node: SyntaxNode): SnippetPart {
  const raw = node.toText();
  let quotes = 0;
  let content = "";
  for (const element of node.descendantsWithTokens()) {
    if (element.type === "token") {
      if (element.kind === "Quote") {
        quotes += 1;
      } else if (element.kind === "Text") {
        content += element.text;
      }
    }
  }
  return {
    type: "quoted",
    raw,
    decoded: content.replaceAll('\\"', '"'),
    terminated: quotes >= 2,
  };
}

function capturePart(node: SyntaxNode): SnippetPart {
  let text = "";
  let terminated = false;
  for (const element of node.children) {
    if (element.type === "token") {
      if (element.kind === "Text") {
        text += element.text;
      } else if (element.kind === "RBrace") {
        terminated = true;
      }
    }
  }
  return { type: "capture", text, terminated };
}
